cppcache/src/ThinClientLocatorHelper.hpp

```diff
@@ -60,20 +63,23 @@ class ThinClientLocatorHelper {
       std::vector<std::shared_ptr<ServerLocation> >& servers,
       const std::string& serverGrp);
   int32_t getCurLocatorsNum() {
-    return static_cast<int32_t>(m_locHostPort.size());
+    return static_cast<int32_t>(m_Locators.size());
   }
   GfErrType updateLocators(const std::string& serverGrp = "");
 
  private:
+  std::vector<ServerLocation> getLocators() const;
   Connector* createConnection(Connector*& conn, const char* hostname,
                               int32_t port,
                               std::chrono::microseconds waitSeconds,
                               int32_t maxBuffSizePool = 0);
-  std::mutex m_locatorLock;
-  std::vector<ServerLocation> m_locHostPort;
+
+  /**
+   * Data members
+   */
+  mutable std::mutex m_locatorLock;
```

Comment:
The `mutable` keyword is usually a "code smell" - is this just needed in order to declare `getLocators` const?

Reply:
> The `mutable` keyword is usually a "code smell" - is this just needed in order to declare `getLocators` const?

As stated above I don't agree on getLocators no being a const function member.
With this premise, as the object state is not changed per-se, and you are using a mutex as a tool to guarantee mutual exclusion, it would make sense for the mutex to be mutable, so lock/unlock operations are allows in a const-qualified scope.
Regarding this very case, I've seen Herb Sutter has a "M&M rule" which basically summarizes to "mutexes are to be mutable", you can see that here: https://herbsutter.com/2013/05/24/gotw-6a-const-correctness-part-1-3/
Also, I am awaiting on the response for other C++ comitee members opinion on this matter.
I get your point about the code-smell, but in this case marking it as such, it doesn't feel right

Reply:
Also, after seeing your comment it came out to my mind that this very mutex is a good candidate to be a shared_mutex, don't you think? I mean, there are lots of readers, and just a single writer :)

Reply:
shared_mutex seems like a good call, from my reading of it.  Or go with what you have, if you prefer a minimal change.

Just to be clear with respect to getLocators, I wasn't suggesting we remove the const qualifier from the method.  I was just curious if that is why the mutex needed to be mutable.  Your answer is perfectly fine and quite thorough, so leave the mutex variable as mutable, and I'll resolve this conversation.

Reply:
`mutable` on mutex, and the like, is not a code smell but actually an appropriate pattern. The `mutable` her indicates that this, and only this, value can mutate in an otherwise immutable object and that is correct because the mutex is not part of the state of the object but more metadata protecting this state.